### ISSUE TITLE
Support for generic sender parameter in EventHandler

### ISSUE THREAD

**Issue by simplexidev** [dotnet/runtime#28289]
When using event handlers `EventHandler` and `EventHandler<TEventArgs>`, it is not possible to pass a generic sender to the delegate without writing your own delegate type. There should be a event handler that has a second generic parameter to specify the type of sender.

## Rationale/Usage

This is a nice-to-have that would reduce the number of custom event handers (that take just a sender and args) by exposing a new delegate, such as `EventHandler<TSender, TEventArgs>`.

For example, say I had a class with an event that was specific to it and it's derivatives:

```csharp
public delegate void LoadingEventHandler(FooObject sender, EventArgs e);

public class FooObject
{
    public FooObject()
    {
        Loading.Invoke(this, EventArgs.Empty);
    }
    public event LoadingEventHandler Loading;
}
```

With a `EventHandler<TSender, TEventArgs>` delegate, the above could turn into:

```csharp
public class FooObject
{
    public FooObject()
    {
        Loading.Invoke(this, EventArgs.Empty);
    }
    public event EventHandler<FooObject, EventArgs> Loading;
}
```

## Proposed API

```csharp
namespace System
{
    /// <summary>
    /// Represents the method that will handle an event when the event provides data.
    /// </summary>
    /// <typeparam name="TSender">The type of sender object.</typeparam>
    /// <typeparam name="TEventArgs">The type of the event data generated by the event.</typeparam>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">An object that contains the event data.</param>
    public delegate void EventHandler<TSender, TEventArgs>(TSender sender, TEventArgs e)
        where TSender : object;
}
```

## Updates

 - Updated this issue to follow the API review process/other guidelines.
 - Updated the Proposed API with type parameters.
 - Updated the Proposed API: Removed type parameter constraints to align with the existing `EventHandler` implementations, as `TEventArgs` does not have to be of type `EventArgs`. The type constraint for `TSender` remains as `object`.

**Comment by maxkatz6** [comment]
Any updates on the issue?

**Comment by tacdevel** [comment]
Whether we bring over`TypedEventHandler` or have a new delegate for `EventHandler`, I would be happy with it. Are there any further updates on this issue?

**Comment by tom-corwin** [comment]
I'm just checking up on this, as I think it would still be a great addition to the framework.

**Comment by jeffschwMSFT** [comment]
cc @jeffhandley

**Comment by tom-corwin** [comment]
Bumping this again, showing that there's still an interest, although small, in having  this addition.

**Comment by init-center** [comment]
+1

**Comment by simplexidev** [comment]
Guess it's been over a year since I bumped this. So Hi again. Still have interest

**Comment by hawkerm** [comment]
Yeah, just been wondering this same thing myself. Would be nice if there was a nice generic event type for this, even with a different name like `TypedEventHandler`. I'm always casting the object, it's very tiring when I know what the sender is within my own code.

**Comment by Sergio0694** [comment]
> "instead of custom delegates, when defining APIs with callbacks"

That doesn't seem like it applies to event handlers at all, but rather to APIs like `Process<T>(T[] items, Action<T> func)`.

---

Having `EventHandler<TSender, TArgs>` would also be particularly useful for WinRT, which has [`TypedEventHandler<T, TArgs>`](https://learn.microsoft.com/en-us/uwp/api/windows.foundation.typedeventhandler-2?view=winrt-26100). This has always been a point of friction, with that delegate not being part of the normal .NET BCL. If we added `EventHandler<TSender, TArgs>`, we could finally update the projections to just use it (and hide `TypedEventHandler<,>` entirely), which would make things much more consistent with `EventHandler<TArgs>`, and remove issues especially when having multiple projects targeting different TFMs.

**Comment by NinoFloris** [comment]
TypedEventHandler is violating the object sender guideline.

Personally I do just use Action that takes EventArgs, apparently violating the System.EventHandler<T> guideline. There isn't anything special about those event delegates, and the convention probably stems from the time delegates were scary.

IMO it's all whatever, as long as you're consistent.

**Comment by hawkerm** [comment]
Do we know why the framework guidance says the sender should be `object`? It seems like a holdover from when many events were overloaded in the past and called from many different places. But in many cases when creating events, you're only ever raising it from a single place (though this is anecdotal on my part). I'm sure data could be collected on EventHandler usage on GitHub to determine how frequently an event is overloaded and called without the direct context of its sending class.

And, as much as I love the [framework design guidelines](https://learn.microsoft.com/dotnet/standard/design-guidelines/) too, they're not kept up to date on the docs site for some reason:

> These guidelines are excerpted from the book Framework Design Guidelines: Conventions, Idioms, and Patterns for Reusable .NET Libraries, 2nd Edition, by Krzysztof Cwalina and Brad Abrams, which was published in 2008. The book has since been fully revised in the [third edition](https://www.informit.com/store/framework-design-guidelines-conventions-idioms-and-9780135896464). _Some of the information in these guidelines may be out-of-date._

It'd be nice if whatever deal was made to publish them openly on the docs could be renewed? Or if they could be updated to align to whatever latest practices the .NET team would advise for developers today.

**Comment by tannergooding** [comment]
After an offline discussion with @Sergio0694, @bartonjs, and @KrzysztofCwalina we're going to take this to API review.

We should be prepared for lengthy discussion about the design guideline before ever getting to the actual proposal here. If we decide to classify the guideline as outdated we will get to the actual proposal here.

WinForms should be invited to the review.

**Comment by Sergio0694** [comment]
Synced with @tannergooding offline, I'm going to mark this as blocking, as it is so for CsWinRT 3.0 (see #114179). For context: we want to take a dependency on `EventHandler<TSender, TEventArgs>` in the new WinRT projections (replacing [`TypedEventHandler<TSender, TResult>`](https://learn.microsoft.com/en-us/uwp/api/windows.foundation.typedeventhandler-2)), and we can only afford one breaking change in the projection API surface, meaning either we can get this in .NET 10 so we can leverage it in our CsWinRT 3.0 release, or if it slips and we end up releasing before that, we won't be able to ever use it, even if it does make it into a future .NET release. Tanner and I will be at //Build two weeks from now, so it'd be better if we could review this proposal next week, to avoid things shifting too far ahead and close to the deadline for .NET 10 🙂

**Comment by bartonjs** [comment]
* We added contravariant markers
* We removed the constraint on TSender
* We added allows ref struct
* We applied the same changes to the existing EventHandler<T>

```C#
namespace System
{
    public delegate void EventHandler<in TSender, in TEventArgs>(TSender sender, TEventArgs e)
        where TSender : allows ref struct
        where TEventArgs : allows ref struct;
}
```

```diff
namespace System
{
-   public delegate void EventHandler<TEventArgs>(object sender, TEventArgs e);
+   public delegate void EventHandler<in TEventArgs>(object sender, TEventArgs e)
+       where TEventArgs : allows ref struct;
}
```
